org.eclipse.jdt.ls.core/gradle/apt/init.gradle

```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+import org.gradle.api.Project
+import org.gradle.api.Plugin
+import org.gradle.api.file.FileCollection
+import org.gradle.api.plugins.PluginContainer
+import org.gradle.api.tasks.compile.CompileOptions
+import org.gradle.api.tasks.compile.JavaCompile
+import org.gradle.tooling.provider.model.ToolingModelBuilder
+import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
+import org.gradle.util.GradleVersion
+
+import javax.inject.Inject
+
+class GradleAnnotationProcessorPatchPlugin implements Plugin<Project> {
+
+	private final ToolingModelBuilderRegistry registry
+
+	@Inject
+	GradleAnnotationProcessorPatchPlugin(ToolingModelBuilderRegistry registry) {
+		this.registry = registry
+	}
+
+	@Override
+	void apply(Project project) {
+		this.registry.register(new AnnotationProcessorModelBuilder())
+	}
+
+	private static class AnnotationProcessorModelBuilder implements ToolingModelBuilder {
+
+		@Override
+		boolean canBuild(String modelName) {
+			return "java.util.Map" == modelName
+		}
+
+		@Override
+		Object buildAll(String modelName, Project rootProject) {
+			final GradleVersion current = GradleVersion.current().getBaseVersion()
+			if (current < GradleVersion.version("5.2")) {
+				return Collections.emptyMap()
+			}
+
+			Set<Project> allProject = rootProject.getAllprojects()
+			Map<File, Map<String, Object>> annotationProcessorInfo = new HashMap<>()
+			for (Project project : allProject) {
+				PluginContainer plugins = project.getPlugins()
+				if (!hasPlugin(plugins, "java")) {
+					continue
+				}
+
+				Set<File> processors = new HashSet<>()
+				List<String> compilerArgs = new LinkedList<>()
+				// Compiler args for test are ignored.
+				// Due to JDT does not differentiate the args between main and test.
+				collectApConfiguration(project, "compileJava", processors, compilerArgs)
```

Comment:
There may be different processors present on `compileJava` vs `compileTestJava`, how does this PR address that?

Reply:
The reason that test annotation processors are ignored is that the APT support on JDT side cannot differentiate the processors from different source sets. My approach here is to ignore test annotation processors.

Another approach we can take is to mix all the processors from main and test together, but personally I think this is not ideal neither.

This is the case that we cannot handle gracefully now until the JDT can support that.

Reply:
I filed an issue to track the `testAnnotationProcessor` support: https://github.com/eclipse/eclipse.jdt.ls/issues/2345, which we can address separately from this pull request.